packages/relay-compiler/codegen/writeRelayGeneratedFile.js

```diff
@@ -95,6 +99,12 @@ async function writeRelayGeneratedFile(
       hash = hasher.digest('hex');
       return extractHash(oldContent);
     });
+
+    const {text} = generatedNode.params;
+    if (persistQuery && text != null) {
+      await persistQuery(text, sourceHash, filename);
+    }
+
```

Comment:
it's being discarded because the sourceHash that is passed into persistQuery at the moment https://github.com/facebook/relay/blob/master/packages/relay-compiler/codegen/writeRelayGeneratedFile.js#L118
is just returned here https://github.com/facebook/relay/blob/master/packages/relay-compiler/bin/RelayCompilerMain.js#L310 so I just set it directly, here https://github.com/facebook/relay/pull/2629/files#diff-7105a5c68c6256bab3fd91e13e7d65c9R128

Surely if I move it below the persistQuery branch it won't be hit, since it'll return here https://github.com/facebook/relay/pull/2629/files#diff-7105a5c68c6256bab3fd91e13e7d65c9R110
Unless I've misunderstood?

Reply:
Ah - sorry, that is a reasonable assumption given just the code visible in OSS. You're correct that the *default* implementation of `persistQuery` just returns the input, but our internal implementation actually persists the query and returns a persisted id that may differ from source hash. But we also don't need the query map, so it's fine to build up a sourceHash => text mapping for the query map, so long as we keep setting the `id` on the generated node to the output of the `persistQuery` function.

Reply:
Ah got it. You're moving the call to `persistQuery` here to ensure that it's called for all queries, even those that don't change, in order to have a complete query map in watch mode. Note that our internal `persistQuery` has to actually persist the query so we want to avoid calling it unless it's necessary. I think we need to split up the responsibilities so that there's one "registerQuery" or similar that is always called at this point just to save the id/text mapping, and then the original `persistQuery` called below, only when there are changes.

Reply:
Aha! Ok, I think I've got you. I added some comments to help others if they're confused by that function, but I don't know if that's ok or if what I put was accurate. I added a registerQuery function to the WriterConfig and moved the actual setting of the map entry into that.

Reply:
> Ah - sorry, that is a reasonable assumption given just the code visible in OSS. You're correct that the _default_ implementation of `persistQuery` just returns the input, but our internal implementation actually persists the query and returns a persisted id that may differ from source hash.

Not sure how important this is so just adding for context, but iirc @robrichard does this too in their setup.